gh-130599: precompute conversion constants for ``long()``

nascheme opened the pull request:
This avoids a data race in free-threaded builds due to mutating the static conversion arrays at runtime.  Instead, compute and initialize the constants at runtime initialization.  We only need to do this once (not once per interpreter) and so I added `_PyLong_InitRuntime()`.


<!-- gh-issue-number: gh-130599 -->
* Issue: gh-130599
<!-- /gh-issue-number -->

nascheme committed:
gh-130599: precompute conversion constants for ``long()``

This avoids a data race in free-threaded builds due to mutating the
statically arrays at runtime.  Instead, compute and initialize the constants
at runtime initialization.

kumaraditya303 commented on Objects/longobject.c:
```diff
@@ -2854,6 +2849,21 @@ long_from_non_binary_base(const char *start, const char *end, Py_ssize_t digits,
+    PyLongObject *z;
+    const char *p;
+
+    assert (log_base_BASE[base] != 0.0); // pre-computed by _PyLong_InitRuntime()
```
```suggestion
    assert(log_base_BASE[base] != 0.0); // pre-computed by _PyLong_InitRuntime()
```

nascheme committed:
Code tidy for assert().

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

nascheme committed:
Update c-analyzer/cpython/ignored.tsv.

nascheme committed:
Use static tables rather than computing on startup.

nascheme committed:
Remove unneeded func declaration, fix comment.

colesbury commented on Objects/longobject.c:
```diff
@@ -2820,6 +2820,59 @@ that triggers it(!).  Instead the code was tested by artificially allocating
+        0.3271260397072346, 0.3302797540257917, 0.33333333333333337,
+        0.3362929412905636, 0.3391641894166893, 0.34195220112966446,
+        0.34466166676282084};
+    static int convmultmax_base[37] = {0, 0, 32768, 19683, 16384,
```
The type of convmultmax_base should be `twodigits`

colesbury commented on Objects/longobject.c:
```diff
@@ -2820,6 +2820,59 @@ that triggers it(!).  Instead the code was tested by artificially allocating
+        2744, 3375, 4096, 4913, 5832, 6859, 8000, 9261, 10648, 12167,
+        13824, 15625, 17576, 19683, 21952, 24389, 27000, 29791, 32768,
+        1089, 1156, 1225, 1296};
+    static twodigits convwidth_base[37] = {0, 0, 15, 9, 7, 6, 5, 5, 5,
```
`convwidth_base` should be `int`

nascheme replied:
Ah, good catch.  I swapped the array order in my Python script but not the types.  Fixed.

colesbury commented on Objects/longobject.c:
```diff
@@ -2820,6 +2820,59 @@ that triggers it(!).  Instead the code was tested by artificially allocating
 digit beyond the first.
 ***/
+
+// Tables are computed by Tools/scripts/long_conv_tables.py
+#if PYLONG_BITS_IN_DIGIT == 15
```
I'd consider putting the generated code in a `.c.h` file and `#include` it here, but up to you.

nascheme replied:
IMHO, it's not quite enough lines to justify a separate file (adds about 50 new lines).

colesbury reviewed (commented):
I measured the difference for `./python -c "pass"` between this and the earlier commit that computed the values at runtime initialization. It's very small, about 0.1%, so I if you prefer that approach, that seems fine to me too.

nascheme committed:
Fix array declarations.

Use "const" and fix the swapped types of convmultmax_base and
convwidth_base.

nascheme committed:
Remove unused 'steps' variable.

nascheme committed:
Merge 'origin/main' into gh-130599-long-conv-ts-fix

nascheme committed:
Don't compute unneeded constants.

This makes the pre-computed tables match exactly what was previously
computed at runtime, if all base conversions were used.  This makes
them slightly smaller too.
